### ISSUE TITLE
Security: Match Kyma-CR-Domain with SAN from Watcher request

### ISSUE THREAD

**Issue by tobiscr** [kyma-project/lifecycle-manager#149]
## Description

The SKR watcher as externally running component has to communicate to the KCP listener endpoints. This could lead to several risks like DOD-attaches or man-in-the-middle attacks.
To authenticate the incoming requests, we will implement a Domain-Name-Pinning.
### Setup
- KEB (Kyma Environment Broker) places the domain name of the prepared SKR cluster inside of the corresponding Kyma CR as annotation - called `skr-domain`
- SKR is deployed with enabled api extension called `Shoot-Cert-Service`, which enables the SKR cluster to request for a trusted X.509 certificate.
- During the Watcher deployment on the SKR cluster, a `certificate.cert.gardener.cloud` custom resource(CR) needs to be created. This CR will then be picked up by the extension-controller of Gardener to request for a private key and a certificate ([more information](https://gardener.cloud/docs/extensions/others/gardener-extension-shoot-cert-service/docs/usage/request_cert/)).
  - Private Key and Certificate are stored in a Secret on the SKR.
  - Name and Namespace of the certificate secret can be obtained of the `certificate` CR: `spec.secretRef.name`, `spec.secretRef.namespace`
- Gardener’s certificate service regularly checks the expiration timestamp of Certificates, triggers a renewal process if necessary and replaces the existing X.509 certificate with a new one. (Always given, if `Shoot-Cert-Service` api extension is enabled)

### Sending a Request from Watcher (SKR) to KCP:
1. Request will be sent from SKR cluster to the KCP Gateway using mTLS, thus the certificate will be sent with the request to the KCP Istio gateway
3. Runtime-Watcher request will be send from the SKR cluster to the Istio Gateway of the KCP cluster using encrypted mTLS. For the mTLS connection the requested private key and certificate will be used.
7. Istio Gateway will receive request. (mTLS Connection handshake etc. will not be mentioned for simplification)
   1. Gateway will set the received certificate as the `x-forwarded-client-cert` (XFCC) proxy header ([more information](https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#configuring-x-forwarded-client-cert-headers)).
   2. Will forward the request with newly set header to the VirtualService.
8. VirtualService will received the request and forward it to the corresponding operator, in this case the `lifecycle-manager`.
9. Lifecycle-Manager received request:
   1. Extracts the SAN(`subject alternative name`) out of the certificate.
   2. Fetches the KymaCR mentioned in the request payload.
   3. Checks if specified domain in the fetched Kyma CR matches with the given SAN
   4. 
      1. Matches: Sender is authorised to trigger a reconciliation for the stated KymaCR in the request body &#8594; Reconciliation will be trigger.
      2. Does NOT match: Sender is not authorised to trigger a reconciliation for the stated KymaCR of the request body &#8594; Reconciliation will not be trigger and request will be dropped.

**ACs:**
- [ ] Configure KCP gateway to set the received certificate as the `x-forwarded-client-cert` (XFCC) proxy header
- [ ] Lifecycle-Manager checks if KymaCR from received payload contains the same domain name as the SAN

**Depends on** (full story):
https://github.com/kyma-project/lifecycle-manager/issues/289
https://github.com/orgs/kyma-project/projects/30/views/12?filterQuery=-board%3APipeline+-board%3ASecurity+-status%3ADone%2CEPIC+refined%3ARefined+mtls

## Steps to exploit
Deploy a second modified watcher in the SKR cluster. The watcher normally has the KymaName included in the request body, which needs to be reconciled. But this modified watcher will include a different KymaName from another SKR in the request body. Thus, this modified watcher would be able to trigger a reconciliation for a different SKR cluster, without having the correct access.

**Comment by jeremyharisch** [comment]
Proposed workflow:
![sign_watcher_request drawio](https://user-images.githubusercontent.com/48282931/196878914-cd846c52-8c07-4aab-9adc-25b1d01fe39c.svg)

**Comment by jeremyharisch** [comment]
Different Proposal:
// TODO - Workflow diagram

- Validation webhook checks for private Key, stored in a secret
    - If yes, then use private Key to sign the request and send it to KCP
    - If no, then send a request to a `key-request` endpoint of KCP. KCP generates new Key pair as proposed in previous comment.

Problem:
Anyone can request a new Key.

**Comment by jakobmoellersap** [comment]
I have a few questions here:

### Sending a Request from Watcher (SKR) to KCP:
1. Request will be signed using Private Key stored in Secret
=> the private key secret could be reused by anyone within the same cluster, is this known @TorstenD-SAP 
3. If Private Key does not exists
   a. Log error
   b. Do not send request
4. Signed request will be send to the Gateway of the KCP using encrypted mTLS
5. VirtualService will forward already decrypted request to lifecycle-manager
=> it will not send the request fully decrypted but again encrypted with istio (inside service mesh we also have mtls), why cant we just forward the MTLs Certificate from the gateway and use this for the verification that the cluster is authorized to trigger a reconciliation (with XFF headers in the clusters)? @TorstenD-SAP 
7. Lifecycle-Manager checks the body of the request and fetches the corresponding KymaCR
8. Using the fetched KymaCR, the corresponding Public Key will be fetched
9. Using this Public Key, the signed request will be verified
  a. If request could be verified: Sender is authorised to trigger a reconciliation for the stated KymaCR in the request body &#8594; Reconciliation will be trigger 
  b. If request could NOT be verified: Sender is not authorised to trigger a reconciliation for the stated KymaCR of the request body &#8594; Reconciliation will not be trigger and request will be dropped

TLDR: I think we could simply reuse the MTLs certificate if its forwarded via header from the gateway, instead of proxying it every time for requests, especially since we anyhow want to use the same keys.

**Comment by TorstenD-SAP** [comment]
> the private key secret could be reused by anyone within the same cluster

This is something which cannot be avoided because we have to store this private key somewhere in the SKR cluster of the customer. We should ensure, that by default the secret is secured and can only be used by the watcher.

> it will not send the request fully decrypted but again encrypted with istio (inside service mesh we also have mtls), why cant we just forward the MTLs Certificate from the gateway and use this for the verification that the cluster is authorized to trigger a reconciliation (with XFF headers in the clusters)? @TorstenD-SAP I think this was recently added in the Istio component with X-Forwarded-Cert

I understand the comment in that way, that the idea is to forward the SKR clusters certificate, which is used by the Gateway to establish the mTLS connection with the watcher on the SKR cluster, from the Gateway to the lifecycle manager by using the  X-Forwarded-Client-Cert header (https://istio.io/latest/docs/ops/configuration/traffic-management/network-topologies/#configuring-x-forwarded-client-cert-headers). This certificate can then be used by the lifecycle manager to uniquely identify the SKR cluster. From my point, it is not enough to just check if the SKR cluster is authorized to trigger a reconciliation. We also have to ensure, that only the authorized SKR cluster is reconciled by this request in order to ensure, that the reconciliation request cannot be misused to reconcile other SKR clusters. But if we can ensure this and the used keys and certificates are unique for each SKR cluster, it should be possible to use the existing keys and certificates which are used for mTLS too.

**Comment by jeremyharisch** [comment]
Ticket description has been updated with the agreed solution

**Comment by jeremyharisch** [comment]
To introduce this feature two PRs have been opened -  Please review in stated order:
1. https://github.com/kyma-project/runtime-watcher/pull/89 
2. https://github.com/kyma-project/lifecycle-manager/pull/321

If you need help testing, please ping me - I can help with setting up a cluster with needed certificates, VirtualService, etc.

By default the Domain-Name-Pinning has been turned off, but can be enabled using a flag.
